yt_dlp/extractor/zdf.py

```diff
@@ -410,32 +413,72 @@ def _og_search_title(self, webpage, fatal=False):
         title = super()._og_search_title(webpage, fatal=fatal)
         return re.split(r'\s+[-|]\s+ZDF(?:mediathek)?$', title or '')[0] or None
 
+    def _extract_document_id(self, webpage):
+        matches = re.search(r'docId\s*:\s*[\'"](?P<docid>[^\'"]+)[\'"]', webpage)
+        return matches and matches.group('docid')
+
+    def _get_playlist_description(self, page_data):
+        headline = traverse_obj(page_data, ('shortText', 'headline'))
+        text = traverse_obj(page_data, ('shortText', 'text'))
+        if headline is not None and text is not None:
+            return f'{headline}\n\n{text}'
+        return headline or text
+
+    def _convert_thumbnails(self, thumbnails):
+        return [{
+            'id': key,
+            'url': thumbnail_info['url'],
+            'width': int_or_none(thumbnail_info.get('width')),
+            'height': int_or_none(thumbnail_info.get('height')),
+        } for key, thumbnail_info in thumbnails.items() if url_or_none(thumbnail_info.get('url'))]
+
+    def _teaser_to_url_result(self, teaser):
+        return self.url_result(
+            teaser['sharingUrl'], ie=ZDFIE.ie_key(),
+            id=teaser.get('id'), title=teaser.get('titel', ''),
+            thumbnails=self._convert_thumbnails(teaser.get('teaserBild', {})),
+            description=teaser.get('beschreibung'),
+            duration=float_or_none(teaser.get('length')),
+            media_type=teaser.get('currentVideoType') or teaser.get('contentType'),
+            season_number=int_or_none(teaser.get('seasonNumber')),
+            episode_number=int_or_none(teaser.get('episodeNumber')))
+
     def _real_extract(self, url):
         channel_id = self._match_id(url)
 
         webpage = self._download_webpage(url, channel_id)
 
-        matches = re.finditer(
-            rf'''<div\b[^>]*?\sdata-plusbar-id\s*=\s*(["'])(?P<p_id>[\w-]+)\1[^>]*?\sdata-plusbar-url=\1(?P<url>{ZDFIE._VALID_URL})\1''',
-            webpage)
+        main_video = None
+        playlist_videos = []
+
+        document_id = self._extract_document_id(webpage)
+        if document_id is not None:
+            data = self._download_mediathekv2_document(document_id)
+
+            for cluster in data['cluster']:
+                for teaser in cluster['teaser']:
+                    if cluster['type'] == 'teaserContent' and teaser['type'] == 'video':
+                        main_video = main_video or teaser
+                    elif cluster['type'] == 'teaser' and teaser['type'] == 'video':
+                        if teaser['brandId'] != document_id:
+                            # These are unrelated 'You might also like' videos, filter them out
+                            continue
+                        playlist_videos.append(teaser)
 
         if self._downloader.params.get('noplaylist', False):
```

Comment:
I know the old code was doing this, but we really should be using `_yes_playlist()` instead

Reply:
Yeah, I was thinking that this looked weird when I saw it.